lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -262,18 +262,20 @@ SetOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
                         OpaqueProperties properties, RegionRange regions,
                         llvm::SmallVectorImpl<Type> &inferredReturnTypes) {
 
-  Value leftInput = operands[0];
-  Value rightInput = operands[1];
-
-  TypeRange leftFieldTypes = cast<TupleType>(leftInput.getType()).getTypes();
-  TypeRange rightFieldTypes = cast<TupleType>(rightInput.getType()).getTypes();
+  ValueRange inputs = operands;
 
-  if (leftFieldTypes != rightFieldTypes)
-    return ::emitError(loc.value())
-           << "left and right inputs must have the same field types";
+  if (inputs.size() < 2)
+    return ::emitError(loc.value()) << "expected at least 2 inputs";
 
-  auto resultType = TupleType::get(context, leftFieldTypes);
+  TypeRange fieldType = cast<TupleType>(inputs[0].getType()).getTypes();
 
+  for (Value input : inputs) {
+    TypeRange inputFieldTypes = cast<TupleType>(input.getType()).getTypes();
+    if (fieldType != inputFieldTypes)
+      return ::emitError(loc.value())
+             << "all inputs must have the same field types";
+  }
+  auto resultType = TupleType::get(context, fieldType);
```

Comment:
Nit: `TupleType::get` has non-trivial costs because it needs to look up the newly generated type in the context. (There is always only one instance of any possible type such that we can compare types by comparing the pointers those types.) Currently, there is a cheaper alternative, namely `inputs[0].getType()` (which should live in a dedicated variable where used first).